### DISCUSSION TITLE
Results are very bad on trained model

### DISCUSSION THREAD

**Discussion by khawar-islam** [mindee/doctr#1167] [Q&A]
Dear @felixdittrich92 @frgfm 

I am very confused and I have no idea what's going on? I have trained a MASTER model on 10M hangul images and If you look the validation set the accuracy is very good and it works fine on validation set. FYI, I did not provide the `--font` argument in my training script and of course this is not the case of model failure because during validation set it works fine

**Training Script  Changes**
`    vocab = "".join(OrderedDict.fromkeys(VOCABS['korean']))`

**Training Log**
```
Train set loaded in 48.33s (7769435 samples in 47374 batches)
Validation loss decreased inf --> 0.00745843: saving state...                                                                                                   
Epoch 1/10 - Validation loss: 0.00745843 (Exact: 96.62% | Partial: 96.62%)
Validation loss decreased 0.00745843 --> 0.00339667: saving state...                                                                                            
Epoch 2/10 - Validation loss: 0.00339667 (Exact: 98.33% | Partial: 98.33%)
Validation loss decreased 0.00339667 --> 0.00228179: saving state...                                                                                            
Epoch 3/10 - Validation loss: 0.00228179 (Exact: 98.87% | Partial: 98.87%)
Validation loss decreased 0.00228179 --> 0.00174915: saving state...                                                                                            
Epoch 4/10 - Validation loss: 0.00174915 (Exact: 99.13% | Partial: 99.13%)
Validation loss decreased 0.00174915 --> 0.00128966: saving state...                                                                                            
Epoch 5/10 - Validation loss: 0.00128966 (Exact: 99.32% | Partial: 99.32%)
Validation loss decreased 0.00128966 --> 0.00101758: saving state...                                                                                            
Epoch 6/10 - Validation loss: 0.00101758 (Exact: 99.46% | Partial: 99.46%)
Validation loss decreased 0.00101758 --> 0.000899193: saving state...                                                                                           
Epoch 7/10 - Validation loss: 0.000899193 (Exact: 99.52% | Partial: 99.52%)
Validation loss decreased 0.000899193 --> 0.000809529: saving state...                                                                                          
Epoch 8/10 - Validation loss: 0.000809529 (Exact: 99.55% | Partial: 99.55%)
Validation loss decreased 0.000809529 --> 0.000769762: saving state...                                                                                          
Epoch 9/10 - Validation loss: 0.000769762 (Exact: 99.57% | Partial: 99.57%)
Validation loss decreased 0.000769762 --> 0.000752969: saving state...                                                                                          
Epoch 10/10 - Validation loss: 0.000752969 (Exact: 99.58% | Partial: 99.58%)
```
Now, I have tested  on several images which are different from validation set
![test60](https://user-images.githubusercontent.com/11488932/225793751-41b1cbf6-308e-4fa4-b55d-ed19fa710aa6.jpeg)

**Output** (Completely wrong)
`[('두리번거리다', 0.5222912430763245)]`

**Comment by frgfm** [comment]
Hi @khawar-islam :wave:

Glad to hear people are trying that on other vocabularies!
So my suspicion is that there is a problem with the encoding of the hangul characters when rendering the synthetic dataset, or simply that the distribution is too far from reality (the way the characters are rendered with that font doesn't have enough intra-class variability)

I suggest two things:
- checking that the synthetic trained images match your visual expectation: try the `--show-samples` argument, it should help you with that visual inspection
- if there is no problem there, it will be a bit more manual, but I would check the validation ones

My intuition here is that because of the apparent validation metrics being quite good, the problem would be on the first part (that several characters cannot be rendered correctly which results in different hangul characters being rendered as another "error" character).

I hope this helps, let us know how it goes!

**Reply by khawar-islam** [reply]
Hello @frgfm 

I have made a validation dataset myself and I just visually inspect the dataset and it shows **??** why? I can visualize the dataset in my folder and there are only 20 images.

`python3 train_pytorch.py master --val_path /media/cvpr/CM_22/KR_2M/val --show-samples
`
![image](https://user-images.githubusercontent.com/11488932/226149439-7b8acf1b-6e08-45de-8671-0df26dfe918f.png)

**In Folder**

![image](https://user-images.githubusercontent.com/11488932/226149491-0e9f9cec-0fd5-4870-858f-87c0f0af03f9.png)

**Reply by frgfm** [reply]
Show samples show you the train samples. When you provide a path, it uses your data, otherwise it's generated. And as you can see on your first screenshot, the characters are not rendered properly.

As I mentioned above, if you don't specify the font, it will pick some by default. And it looks like those fonts don't support the characters in your vocabulary. So I suggest to either :
- download a font that support those and check if it renders correctly
- or pass the path to a folder with training samples
Hope this helps!

**Reply by khawar-islam** [reply]
Hello, @frgfm I have included all fonts `--fonts` parameters and now I can visualize correctly still if you can see the training log. Why `partial is 0.00%`?

**Training Log**
![image](https://user-images.githubusercontent.com/11488932/226490277-6393e552-e90f-4e20-a8f0-403cee6755c0.png)

**Reply by frgfm** [reply]
Well, hard for me to debug the training loop from someone else :sweat_smile: 
But it's good if when you visualize the synthetic data, it now looks good. I doubt this is the case but perhaps the architecture is not good enough for this. You could check this because doing a character classification quick training loop (and you'll be able to use the background afterwards). I suggest checking the reference script for this

If character classification works well, then there is an issue somewhere in your text recognition training loop :upside_down_face:

**Reply by felixdittrich92** [reply]
Hi @khawar-islam 👋🏼 ,
I thought we were on the right track by now !? 😅 
ref.: https://github.com/mindee/doctr/discussions/1103#discussioncomment-4152420

Btw. have you tested the vitstr architecure trained by https://github.com/mindee/doctr/issues/1133 (maybe a good fit also for you ?)
https://huggingface.co/Bingsu/vitstr_small-korean

**Reply by khawar-islam** [reply]
@felixdittrich92 Can you please check your email? my bad, yes I remember we are on the right track. I have decided to include English and digits datasets but due to the MultiGPU training problem, I am completely stuck and still stuck.

Last time, my KR dataset (2M) is not good and there are lots of distorted images (Even humans cannot understand training images). 

After that I played some synthtiger parameters and generate a more good dataset now I am training new model with 10M images but literally, now I failed I have never changed anything in the code.

#1133 is not a good idea and their results are very poor. Because a single KR word consists of multiple characters and there are 11 thousand possible vocabularies to cover KR language.

![](https://camo.githubusercontent.com/fa06fefe28672dc3abae56fda3cf57c5a4c8055514ee87e47782a8d2a650a110/68747470733a2f2f692e696d6775722e636f6d2f4b7a4b713857592e6a7067)

**Reply by khawar-islam** [reply]
@frgfm Can you please check your slack?

**Reply by khawar-islam** [reply]
Dear @frgfm please have a look at slack

**Reply by khawar-islam** [reply]
Hello @frgfm @felixdittrich92  I have just changed the `vocab = "".join(OrderedDict.fromkeys(VOCABS['korean'])) ` in the `classification.pytorch.py` file and  trained the classification model, here are the results

`python references/classification/train_pytorch.py mobilenet_v3_large --epochs 2 --device 0 --font "ko/NanumBarunGothicBold.ttf,ko/나눔손글씨 따악단단.ttf,ko/NanumBarunGothicLight.ttf,ko/나눔손글씨 딸에게 엄마가.ttf,ko/NanumBarunGothic.ttf,ko/나눔손글씨 또박또 박.ttf"`

**Epochs Results**
```
Epoch 1/5 - Validation loss: 0.0667947 (Acc: 98.01%)
Epoch 2/5 - Validation loss: 0.0142913 (Acc: 99.57%)
Epoch 3/5 - Validation loss: 0.00275385 (Acc: 99.92%)
Epoch 4/5 - Validation loss: 0.000653031 (Acc: 99.97%)
Epoch 5/5 - Validation loss: 0.000455066 (Acc: 99.98%)
```

**Reply by frgfm** [reply]
Hi @khawar-islam,
Just FYI: this is an open-source repo, we're happy to help with your inconveniences but we cannot guarantee a SLA/customer service level. As you can imagine, we do have other matters to attend to on a regular week :/

I understand this can be frustrating when you get stuck and maintainers aren't available enough to help quickly, but that's why Felix & I try to give as much information and investigation leads as possible from the beginning. So that DL practitioners can use those directly or finish the investigation by themselves. Also, it's better to keep discussions here rather than via email or slack, so that the next people looking for help will be able to find it publicly available here :smiley: 

Anyway, back to your problem: from your previous discussions with felix, it looks like you had it working until you changed a few things. So I suggest to take back the investigation from there:
- ensure the training synthetic samples are visually satisfactory in your own eye
- if they are, perhaps check that they are as well in the validation loop (considering the transformations aren't the same)
- if your training is working then, and still have a problem, I assume it's from the way inference is done!

Hope this helps :+1:
